controllers/clustertopology_controller_sync.go

```diff
@@ -16,10 +16,86 @@ limitations under the License.
 
 package controllers
 
-import "context"
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
 
 // reconcileState reconciles the current and desired state of the managed Cluster topology.
+// NOTE: We are assuming all the required objects are provided as input; also, in case of any error,
+// the entire reconcile operation will fail. This might be improved in the future if support for reconciling
+// subset of a topology will be implemented.
 func (r *ClusterTopologyReconciler) reconcileState(ctx context.Context, current, desired *clusterTopologyState) error {
-	// TODO: add reconcile logic
+	// Reconcile desired state of the InfrastructureCluster object.
+	if err := r.reconcileInfrastructureCluster(ctx, current, desired); err != nil {
+		return err
+	}
+
+	// TODO: reconcile other objects in the topology
+
+	return nil
+}
+
+// reconcileInfrastructureCluster reconciles the desired state of the InfrastructureCluster object.
+func (r *ClusterTopologyReconciler) reconcileInfrastructureCluster(ctx context.Context, current, desired *clusterTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// If there is no current InfrastructureCluster object, create it.
+	if current.infrastructureCluster == nil {
+		log.Info("creating", desired.infrastructureCluster.GroupVersionKind().String(), desired.infrastructureCluster.GetName())
+		if err := r.Client.Create(ctx, desired.infrastructureCluster); err != nil {
+			return errors.Wrapf(err, "failed to create the %s object", desired.infrastructureCluster.GetKind())
+		}
+		return nil
+	}
+
+	// Check if the current and desired InfrastructureCluster are compatible.
+	// NOTE: After a cluster is created it is assumed that the reference to the infrastructure cluster should
+	// never change (only the content of the object can eventually change). Thus, we are checking for strict compatibility.
+	if err := checkReferencedObjectsAreStrictlyCompatible(current.infrastructureCluster, desired.infrastructureCluster); err != nil {
+		return err
+	}
+
+	// Check differences between current and desired state, and eventually patch the current object.
+	patchHelper, err := newMergePatchHelper(current.infrastructureCluster, desired.infrastructureCluster, r.Client)
```

Comment:
@killianmuldoon good suggestion.
I have added `reconcileReferencedObject` and also `reconcileReferencedTemplate`; the latter does not really belong to this PR so ideally it should be move to your PR for the control plane or to https://github.com/kubernetes-sigs/cluster-api/pull/5072

@sbueringer PTAL to `reconcileReferencedTemplate`; this would allows us to implement the template rotation logic in one single place

Reply:
@fabriziopandini Looks good overall but I think the compatibility check is not the same for every referenced template so we have to make this parameterizable as well (BootstrapTemplate is different)

Reply:
@fabriziopandini If possible, it would be great to rebase this PR after #5038 has been merged now. And it would also be nice if we can find a way to make the compatibility check parameterizable. I would then rebase my PR on top of yours and verify that the `reconcileReferencedTemplate` works in my use case.

Reply:
@sbueringer done; however, as stated above, we should consider to move reconcileReferencedTemplate in your PR or in @killianmuldoon PR